core/src/chunk.rs

```diff
@@ -59,6 +60,77 @@ pub struct Chunk {
     /// Whether this chunk has been modified since the most recent
     /// call to `check_modified`().
     modified: bool,
+
+    heightmaps: [HeightMap; CHUNK_WIDTH * CHUNK_WIDTH],
+}
+
+#[derive(Clone, Copy, Default)]
+pub struct HeightMap {
+    motion_blocking: u8,
+    motion_blocking_no_leaves: u8,
+    ocean_floor: u8,
+    ocean_floor_wg: u8,
+    world_surface: u8,
+    world_surface_wg: u8,
+}
+
+impl HeightMap {
+    /// The highest block that is solid or contains a fluid.
+    pub fn motion_blocking(self) -> u8 {
+        self.motion_blocking
+    }
+
+    pub fn set_motion_blocking(mut self, motion_blocking: u8) {
```

Comment:
Not sure why you're taking in `mut self`, as this passes ownership of the `HeightMap` to the function and any changes get discarded.